src/Servant/API/ContentTypes.hs

```diff
@@ -0,0 +1,181 @@
+data JavaScript deriving Typeable
+data CSS deriving Typeable
+data PlainText deriving Typeable
+data OctetStream deriving Typeable
+
+-- * Accept class
+
+-- | Instances of 'Accept' represent mimetypes. They are used for matching
+-- against the @Accept@ HTTP header of the request, and for setting the
+-- @Content-Type@ header of the response
+--
+-- Example:
+--
+-- > instance Accept HTML where
+-- >    contentType _ = "text" // "html"
+--
+class Accept ctype where
+    contentType   :: Proxy ctype -> M.MediaType
+
+-- | @text/html;charset=utf-8@
+instance Accept HTML where
+    contentType _ = "text" M.// "html" M./: ("charset", "utf-8")
+
+-- | @application/json;charset=utf-8@
+instance Accept JSON where
+    contentType _ = "application" M.// "json" M./: ("charset", "utf-8")
+
+-- | @application/xml;charset=utf-8@
+instance Accept XML where
+    contentType _ = "application" M.// "xml" M./: ("charset", "utf-8")
+
+-- | @application/javascript;charset=utf-8@
+instance Accept JavaScript where
+    contentType _ = "application" M.// "javascript" M./: ("charset", "utf-8")
+
+-- | @text/css;charset=utf-8@
+instance Accept CSS where
+    contentType _ = "text" M.// "css" M./: ("charset", "utf-8")
+
+-- | @text/plain;charset=utf-8@
+instance Accept PlainText where
+    contentType _ = "text" M.// "plain" M./: ("charset", "utf-8")
+
+-- | @application/octet-stream@
+instance Accept OctetStream where
+    contentType _ = "application" M.// "octet-stream"
+
+newtype AcceptHeader = AcceptHeader BS.ByteString
+    deriving (Eq, Show)
+
+-- * Render (serializing)
+
+-- | Instantiate this class to register a way of serializing a type based
+-- on the @Accept@ header.
+--
+-- Example:
+--
+-- > data MyContentType
+-- >
+-- > instance Accept MyContentType where
+-- >    contentType _ = "example" // "prs.me.mine" /: ("charset", "utf-8")
+-- >
+-- > instance Show a => MimeRender MyContentType where
+-- >    toByteString _ val = pack ("This is MINE! " ++ show val)
+-- >
+-- > type MyAPI = "path" :> Get '[MyContentType] Int
+class Accept ctype => MimeRender ctype a where
+    toByteString  :: Proxy ctype -> a -> ByteString
+
+class AllCTRender list a where
+    -- If the Accept header can be matched, returns (Just) a tuple of the
+    -- Content-Type and response (serialization of @a@ into the appropriate
+    -- mimetype).
+    handleAcceptH :: Proxy list -> AcceptHeader -> a -> Maybe (ByteString, ByteString)
+
+instance ( AllMimeRender ctyps a, IsNonEmpty ctyps
+         ) => AllCTRender ctyps a where
+    handleAcceptH _ (AcceptHeader accept) val = M.mapAcceptMedia lkup accept
+      where pctyps = Proxy :: Proxy ctyps
+            amrs = amr pctyps val
+            lkup = zip (map fst amrs) $ map (\(a,b) -> (cs $ show a, b)) amrs
+
+
+
+
+--------------------------------------------------------------------------
+-- * Unrender
+class Accept ctype => MimeUnrender ctype a where
+    fromByteString :: Proxy ctype -> ByteString -> Maybe a
```

Comment:
I agree with using `Either`, but I'm not sure I understand your `UnrenderError` point. If the mime type is not supported, we'll get a `Nothing :: Maybe (Either String ByteString)` in AllCTUnrender, as opposed to a `Just $ Left str` - so we'll know which one it was.